tests/utils/tools.php

```diff
@@ -1,5 +1,188 @@
 <?php
 
+use MongoDB\Driver\Command;
+use MongoDB\Driver\Manager;
+use MongoDB\Driver\ReadPreference;
+use MongoDB\Driver\Server;
+use MongoDB\Driver\Exception\ConnectionException;
+use MongoDB\Driver\Exception\RuntimeException;
+
+/**
+ * Returns the value of a module row from phpinfo(), or null if it's not found.
+ *
+ * @return string|null
+ */
+function get_module_info($row)
+{
+    ob_start();
+    phpinfo(INFO_MODULES);
+    $info = ob_get_clean();
+
+    $pattern = sprintf('/^%s([\w ]+)$/m', preg_quote($label . ' => '));
+
+    if (preg_match($pattern, $info, $matches) !== 1) {
+        return null;
+    }
+
+    return $matches[1];
+}
+
+/**
+ * Returns the primary server.
+ *
+ * @param string $uri Connection string
+ * @return Server
+ * @throws ConnectionException
+ */
+function get_primary_server($uri)
+{
+    return (new Manager($uri))->selectServer(new ReadPreference('primary'));
+}
+
+/**
+ * Returns the storage engine of the primary server.
+ *
+ * @param string $uri Connection string
+ * @return string
+ * @throws RuntimeException
+  */
+function get_server_storage_engine($uri)
+{
+    $server = get_primary_server($uri);
+    $command = new Command(['serverStatus' => 1]);
+    $cursor = $server->executeCommand('admin', $command);
+
+    return current($cursor->toArray())->storageEngine->name;
+}
+
+/**
+ * Returns the version of the primary server.
+ *
+ * @param string $uri Connection string
+ * @return string
+ * @throws RuntimeException
+  */
+function get_server_version($uri)
+{
+    $server = get_primary_server($uri);
+    $command = new Command(['buildInfo' => 1]);
+    $cursor = $server->executeCommand('admin', $command);
+
+    return current($cursor->toArray())->version;
+}
+
+/**
+ * Checks that the topology is a sharded cluster.
+ *
+ * @param string $uri
+ * @return boolean
+ */
+function is_mongos($uri)
+{
+    return get_primary_server($uri)->getType() !== Server::TYPE_MONGOS;
+}
+
+/**
+ * Checks that the topology is a replica set.
+ *
+ * @param string $uri
+ * @return boolean
+ */
+function is_replica_set($uri)
+{
+    return get_primary_server($uri)->getType() !== Server::TYPE_RS_PRIMARY;
+}
+
+/**
+ * Checks if the connection string uses SSL.
+ *
+ * @param string $uri
+ * @return boolean
+ */
+function is_ssl($uri)
+{
+    return stripos($uri, 'ssl=true') !== false;
+}
+
+/**
+ * Checks that the topology is a standalone.
+ *
+ * @param string $uri
+ * @return boolean
+ */
+function is_standalone($uri)
+{
+    return get_primary_server($uri)->getType() !== Server::TYPE_STANDALONE;
+}
+
+/**
+ * Converts the server type constant to a string.
+ *
+ * @see http://php.net/manual/en/class.mongodb-driver-server.php
+ * @param integer $type
+ * @return string
+ */
+function server_type_as_string($type)
```

Comment:
I'm aware that this function isn't currently used. I created it when the topology-based skip functions were producing more verbose messages (e.g. "we need a replica set but found a standalone"). If I don't find a use for it later, I'll delete this before merging.

Reply:
I'd like to leave this in for now, as it may be useful down the line.